### ISSUE TITLE
Add JetStream persistence support for NATS

### ISSUE THREAD

**Issue by franchb** [ClickHouse/ClickHouse#39459]
PR https://github.com/ClickHouse/ClickHouse/issues/32388 implemented support for NATS. 

NATS has a built-in distributed persistence system called JetStream which enables new functionalities and higher qualities of service on top of the base 'Core NATS' functionalities and qualities of service.

JetStream is built-in to `nats-server` and you only need 1 (or 3 or 5 if you want fault-tolerance against 1 or 2 simultaneous NATS server failures) of your NATS server(s) to be JetStream enabled for it to be available to all the client applications.

[More on JetStream](https://docs.nats.io/nats-concepts/jetstream)

`natc.c` already supports JetStream mode - https://github.com/nats-io/nats.c#jetstream

It would be great to have support of NATS JetStream mode in ClickHouse NATS table engine.

**Comment by pkit** [comment]
I'm not sure how it should be supported from clickhouse?
The only things that you can do with jetstream is to get historical data. Which is not really supported in mat views anyway.

**Comment by andrey-mikhailov** [comment]
Hi! JetStream provides durable subscriptions. https://docs.nats.io/legacy/stan/streaming/durables
> Regular subscriptions remember their position while the client is connected. If the client disconnects the position is lost. Durable subscriptions remember their position even if the client is disconnected.
It would be nice to support JetStream in ClickHouse because if the ClickHouse subscriber is down (e.g. CH server is restarting) we will lose Nats messages.
And it also would be great to have "auto_offset_reset" option like in [Kafka](https://clickhouse.com/docs/en/engines/table-engines/integrations/kafka) engine. For example, options can be
- all - Delivers all messages found in the Stream
- new - Delivers only future messages
- last - Delivers the most recent and all future messages
I have grabbed these options from [Nats console client](https://github.com/nats-io/natscli)
```
nats sub --help
```

**Comment by pkit** [comment]
> Durable subscriptions remember their position even if the client is disconnected.

Looks like a good idea

**Comment by lesha888** [comment]
Support would be great to add than.

But now you can use push consumer for your stream for republishing  message from stream to topic for clickhouse. And messages will pushed only if subscriber(CH) listen.
This is partial solution and you can use then now without jetstrem implementation:
```
nats stream add EVENTS --subjects "EVENTS.*" 
nats consumer add EVENTS CH --filter EVENTS.received  --target EVENTS_FOR_CH --deliver-group event_queue  --ack none --deliver all --replay instant
```
```
CREATE TABLE event_queue
(
    raw String
) ENGINE = NATS SETTINGS
    nats_url = '127.0.0.1:4222',
    nats_queue_group='event_queue',
    nats_subjects = 'EVENTS_FOR_CH',
    nats_format = 'RawBLOB'
;
```

**Comment by perestoronin** [comment]
@alexey-milovidov please include this issue in RoadMap 2023.

**Comment by ahmed-adly-khalil** [comment]
@alexey-milovidov it seems that NATS / JetStream only works in single node, i'm guessing this will work more reliably after #42990 ?

**Comment by vdksystem** [comment]
Any news here?

**Comment by alexey-milovidov** [comment]
No news.

**Comment by alexey-milovidov** [comment]
The main blocker is the low quality of tests: https://github.com/ClickHouse/ClickHouse/issues/39185

**Comment by ericbeland** [comment]
@alexey-milovidov It seems like the main blocker for this is gone. 

It is odd that Clickhouse gets paired with Java/Kafka, when NATS seems a more natural pairing--lightweight, fast, and no JVM.
Clickhouse and Clickhouse keeper provide slim, unbloated, options.

First class NATS support through JetStream would let Clickhouse deliver fully on the dream of a JVM-free lifestyle.  :-)

**Comment by alexey-milovidov** [comment]
Yes, I'm in favor of NATS support, and I agree that it is a more natural pairing for ClickHouse.

**Comment by pankajsoni19** [comment]
Is there some plan on it? currently we do batches in Redis then push to clickhouse. we use nats, so this would help alot

**Comment by marvsman** [comment]
Hay there, any update on this?

**Comment by Nps-rf** [comment]
@dmitry-sles-novikov Can an issue be considered closed according to #84799?

**Comment by dmitry-sles-novikov** [comment]
@Nps-rf yes, you can try NATS JetStream in 25.9
